cogs/utils/json_helpers.py

```diff
@@ -0,0 +1,132 @@
+import os
+import discord
+from collections import defaultdict
+from cogs.utils.dataIO import dataIO
+
+GLOBAL_KEY = '__global__'
+SENTINEL = object()
+
+
+class DataDB:
+    """
+    A helper class to streamline the saving of json files
+    """
+    def __init__(self, file_path, *, create_dirs=False, default_value=SENTINEL):
+        self.path = file_path
+
+        file_exists = os.path.isfile(file_path)
+
+        if create_dirs and not file_exists:
+            path, _ = os.path.split(file_path)
+            if path:
+                try:
+                    os.makedirs(path)
+                except FileExistsError:
+                    pass
+
+        if file_exists:
+            self._data = dataIO.load_json(file_path)
+        else:
+            self._data = {}
+            self._save()
+
+        if default_value is not SENTINEL:
+            def _get_default():
+                return default_value
+            self._data = defaultdict(_get_default, self._data)
+
+    def set(self, key, value):
+        """Sets a DB's entry"""
+        self._data[key] = value
+        self._save()
+
+    def get(self, key, default=None):
+        """Returns a DB's entry"""
+        return self._data.get(key, default)
+
+    def remove(self, key):
+        """Removes a DB's entry"""
+        del self._data[key]
+        self._save()
+
+    def wipe(self):
+        """Wipes DB"""
+        self._data = {}
+        self._save()
+
+    def all(self):
+        """Returns all DB's data"""
+        return self._data
+
+    def _save(self):
+        dataIO.save_json(self.path, self._data)
+
+    def __contains__(self, key):
+        return key in self._data
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+    def __len__(self):
+        return len(self._data)
+
+
+class ServerDB(DataDB):
+    """
+    A helper class to streamline the saving of server based json data
+    """
+
+    def __init__(self, *args, global_defaults={}, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._global_defaults = global_defaults
+
+    def set(self, server, key, value):
+        """Sets a server's entry"""
+        if not isinstance(server, discord.Server):
+            raise TypeError('Can only set server data')
+        if server.id not in self._data:
+            self._data[server.id] = {}
+        self._data[server.id][key] = value
+        self._save()
+
+    def get(self, server, key):
```

Comment:
Yes, it should in case no key is provided.

Also should make this for getting channel data as well if ``channel`` is not None. So I suggest doing something like

```py
    def get(self, server=None, channel=None, key):
         """Returns a server or channel's entry"""
         if channel is not None:
             if not isinstance(channel, discord.Channel):
                 raise TypeError('Can only get channel or server data')
             return self._data[channel.id][key]
         else:
             if server is not None:
                 if not isinstance(server, discord.Server):
                     raise TypeError('Can only get channel or server data')
                 return self._data[server.id][key]
```

So that way people can tell the bot to explicitly do special things per channel on a server if they want to have it do certain things in certain channels on their server(s).